hoot-core/src/main/cpp/hoot/core/io/OsmApiDbAwareHootApiDbWriter.cpp

```diff
@@ -0,0 +1,290 @@
+      {
+        retVal = _nodeRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Node);
+        _nodeRemap[eid.getId()] = retVal;
+        if (_outputMappingFile.length() > 0)
+        {
+          _sourceNodeIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+    case ElementType::Way:
+
+      if (_wayRemap.count(eid.getId()) == 1)
+      {
+        retVal = _wayRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Way);
+        _wayRemap[eid.getId()] = retVal;
+        if ( _outputMappingFile.length() > 0 )
+        {
+          _sourceWayIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+    case ElementType::Relation:
+
+      if (_relationRemap.count(eid.getId()) == 1)
+      {
+        retVal = _relationRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Relation);
+        _relationRemap[eid.getId()] = retVal;
+        if ( _outputMappingFile.length() > 0 )
+        {
+          _sourceRelationIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+  default:
+
+    LOG_ERROR("Tried to create or remap ID for invalid type");
+    throw NotImplementedException();
+
+    break;
+  }
+
+  return retVal;
+}
+
+vector<long> OsmApiDbAwareHootApiDbWriter::_remapNodes(const vector<long>& nids)
+{
+  vector<long> result(nids.size()); // Creates the vector and fills nids.size number of zeroes
+
+  for (size_t i = 0; i < nids.size(); i++)
+  {
+    // This is only called when adding nodes for a way.  If a way has a node we
+    //    did not successfully create a mapping for when importing nodes,
+    //    we can't continue
+    if (_nodeRemap.count(nids[i]) == 1)
+    {
+      result[i] = _nodeRemap.at(nids[i]);
+    }
+    else
+    {
+      //TODO: This isn't good, but haven't figured out how to get around it yet.
+      if (ConfigOptions().getHootapiDbWriterOsmapidbUrl().isEmpty())
+      {
+        throw HootException(QString("Requested ID remap for node " +  QString::number(nids[i])
+          + QString(" but it did not exist in mapping table")));
+      }
+      else
+      {
+        result[i] = nids[i];
+      }
+    }
+  }
+
+  return result;
+}
+
+void OsmApiDbAwareHootApiDbWriter::writePartial(const shared_ptr<const Node>& n)
+{
+  bool countChange = true;
+
+  LOG_VARD(n->getId());
+
+  Tags t = n->getTags();
+  _addElementTags(n, t);
+
+  if (n->getId() < 1)
+  {
+    bool alreadyThere = _nodeRemap.count(n->getId()) != 0;
+    LOG_VARD(alreadyThere);
+    long nodeId = _getRemappedElementId(n->getElementId());
+    LOG_VARD(nodeId);
+    if (alreadyThere)
+    {
+      _hootdb.updateNode(nodeId, n->getY(), n->getX(), n->getVersion() + 1, t);
+    }
+    else
+    {
+      _hootdb.insertNode(nodeId, n->getY(), n->getX(), t);
+    }
+  }
+  else
+  {
+    _hootdb.insertNode(n->getId(), n->getY(), n->getX(), t);
```

Comment:
You mean the hoot api db already containing the element?  You may be right.  All my tests were done writing to a clean hoot api db, so let me come up with a scenario that writes to a hoot api db with data already in it and see what fixes are needed.